### PR TITLE
refactor: Renders Explore in SPA

### PR DESCRIPTION
**Pull request by michael-s-molina**
### SUMMARY
Renders Explore in SPA under `/explore`. Previously, Explore was displayed under `/superset/explore`. This will open a number of possibilities for in-context features and also reduce the development complexity associated with keeping multiple React applications.

This PR also replaces SPA-related Explore navigation with React Router and removes the previous entry point files.

Follow-up of https://github.com/apache/superset/pull/20519

### BEFORE/AFTER SCREENSHOTS OR ANIMATED GIF
https://user-images.githubusercontent.com/70410625/176955027-010c15b7-d3a7-4537-86ef-c8d7e40a6384.mov

### TESTING INSTRUCTIONS
We need to test all features that are associated with Explore, like:
- Navigate from multiple places to Explore (welcome page, charts list, add a chart, dashboard, etc)
- Refresh the Explore page and check if the content persists
- Check permalinks
- Check Explore save flow

### ADDITIONAL INFORMATION
- [ ] Has associated issue:
- [ ] Required feature flags:
- [ ] Changes UI
- [ ] Includes DB Migration (follow approval process in [SIP-59](https://github.com/apache/superset/issues/13351))
  - [ ] Migration is atomic, supports rollback & is backwards-compatible
  - [ ] Confirm DB migration upgrade and downgrade tested
  - [ ] Runtime estimates and downtime expectations provided
- [ ] Introduces new feature or API
- [ ] Removes existing feature or API

### CODE REVIEW

File: superset-frontend/src/explore/exploreUtils/index.js
Comment (kgabryje) [inline superset-frontend/src/explore/exploreUtils/index.js:90]: With this change saving chart won't work until another PR implementing saving chart isn't merged

image

Comment (michael-s-molina) [reply superset-frontend/src/explore/exploreUtils/index.js:90]: You're right. I added a workaround to fix it and a `TODO` to remove the workaround with the next PR.

```
// TODO: This will be removed in the next PR that will change the logic to save a slice
url = url.replace('/explore', '/superset/explore');
```

**Comment by EugeneTorap** [comment]
@michael-s-molina Thank you very much for this long awaited PR
Do you plan to make switching between `Home`, `Dashboard`, `Chart`, `SQL Editor` and `Data` tabs in SPA without reloading the browser?

File: superset/models/slice.py
Comment (kgabryje) [inline superset/models/slice.py:294]: Not sure if `overrides` will be recognized by the new explore endpoint, since it doesn't handle `form_data` param

**Comment by kgabryje** [comment]
We need to replace links to Explore from Datasets List - we still use `<a>` there. However, I'm fine with doing that in a separate PR, since there's an edge case to consider - dataset can link to an external website instead of explore, and for external websites we can't use React Router links. We probably should first create a generic link component, which would detect if url is internal (if so, use React Router link) or external (use HTML anchor)

**Review by kgabryje — APPROVED** [review]
Left 2 non-blocking comments. The PR looks great!

**Review by zhaoyongjie — COMMENTED** [review]
@michael-s-molina The results data pane seems lost.

image

**Comment by kgabryje** [comment]
> @michael-s-molina The results data pane seems lost.

Works for me 🤔 

image

**Comment by zhaoyongjie** [comment]
> > @michael-s-molina The results data pane seems lost.
> 
> Works for me 🤔
> 
> image

Could you try to close the Results Pane and open a new tab?

**Comment by jinghua-qa** [comment]
I have tried open some example charts, i also saw the issue that result/samples are missing, however example chart 'Most Dominant Platforms' is working fine

https://user-images.githubusercontent.com/81597121/177279948-8ffdfa0d-76b2-41a3-8d95-0edd7d90b4f8.mov

**Comment by jinghua-qa** [comment]
Found 2 issue when switching chart:
1, when i switching area chart to pie chart and big number, area chart will non-stop flashing before user update the chart

https://user-images.githubusercontent.com/81597121/177281177-f8da6e33-4d95-47e5-9524-0177f48ea251.mov

2, change chart to pie chart and big number chart through viz switchers, the chart will keep zoom out itself

https://user-images.githubusercontent.com/81597121/177281574-8822faac-e356-439b-aa79-0a0f3c5054f4.mov

**Comment by kgabryje** [comment]
Thanks for spotting that @zhaoyongjie @jinghua-qa ! The problem was that after using the SPA html template, the structure of the page changed a bit and the height of Explore container wasn't calculated correctly. I pushed a fix for that and confirmed that results pane is now displayed correctly when it's closed.

@jinghua-qa I couldn't repro the problem with flashing/zooming, but it looks to me like it's connected to calculating the page height. Hopefully my latest commit fixes this issue as well

**Comment by michael-s-molina** [comment]
> @michael-s-molina Thank you very much for this long awaited PR Do you plan to make switching between `Home`, `Dashboards`, `Charts`, `SQL Editor` and `Data` tabs in SPA without reloading the browser page?

Hi @EugeneTorap. That's the plan for the whole SPA project. We're tackling it in phases so some of the pages may still be reloading.

**Comment by michael-s-molina** [comment]
> Thanks for spotting that @zhaoyongjie @jinghua-qa ! The problem was that after using the SPA html template, the structure of the page changed a bit and the height of Explore container wasn't calculated correctly. I pushed a fix for that and confirmed that results pane is now displayed correctly when it's closed.
> 
> @jinghua-qa I couldn't repro the problem with flashing/zooming, but it looks to me like it's connected to calculating the page height. Hopefully my latest commit fixes this issue as well

Thank you for submitting the fix! I'm just waiting for @zhaoyongjie and @jinghua-qa's approval to merge it.

**Comment by michael-s-molina** [comment]
> We need to replace links to Explore from Datasets List - we still use `<a>` there. However, I'm fine with doing that in a separate PR, since there's an edge case to consider - dataset can link to an external website instead of explore, and for external websites we can't use React Router links. We probably should first create a generic link component, which would detect if url is internal (if so, use React Router link) or external (use HTML anchor)

Good point. Let's tackle it in a separate PR 😉

**Review by zhaoyongjie — APPROVED** [review]

**Review by jinghua-qa — APPROVED** [review]
Verified fix for result/sample title and auto zoom out issue, also did not see major issues for regression test from QA team. LGTM!

**Pull request merged**
